cdpctl/validation/__init__.py

```diff
@@ -57,14 +70,210 @@ class UnrecoverableValidationError(ValidationError):
+
+    @property
+    def render_type(self):
+        """Get the render type."""
+        return self._template.render_type
+
+
+class IssueType(Enum):
+    """Issue Types enum."""
+
+    PROBLEM = "problem"
+    WARNING = "warning"
+
+
+def load_issue_templates(path):
+    """Get the IssueTemplates found in a path."""
+    templates = []
+    schema = IssueTemplateSchema(many=True)
+    with open(os.path.join(path)) as input_file:
+        temp_gen = yaml.load_all(input_file, Loader=yaml.FullLoader)
+        templates = schema.load(temp_gen)
+    return templates
+
+
+def load_all_issue_templates():
+    """Get all of the IssueTemplates found."""
+    issue_templates = {}
+    for root, _, files in os.walk(os.path.dirname(__file__)):
+        if ISSUE_TEMPLATES_FILE in files:
+            loading_templates = load_issue_templates(
+                os.path.join(root, ISSUE_TEMPLATES_FILE)
+            )
+            for template in loading_templates:
+                issue_templates[template.id] = template
+    return issue_templates
+
+
+# List of issues found during the validation run
+_issues: Dict[str, Dict[str, List[Issue]]] = {}
+
+
+def get_issues() -> Dict[str, Dict[str, List[Issue]]]:
+    """Get the Issues found during the validation run."""
+    return _issues
+
+
+_issue_templates: Dict[str, IssueTemplate] = load_all_issue_templates()
+
+
+ValidationContext = namedtuple("CurrentValidationContext", "validation_name")
+
+
+class Context:
+    """Basic Validation Context."""
+
+    def __init__(self) -> None:
+        """Initialize the Context."""
+        self.validation_name = None
+        self.function = None
+        self.nodeid = None
+
+    def clear(self) -> None:
+        """Clear all the context values."""
+        self.validation_name = None
+        self.function = None
+        self.nodeid = None
+
+
+current_context: Context = Context()
+
+
+def validator(func):
+    """Wrap a validator function to handle errors better."""
+
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except Exception as e:
+            if isinstance(e, UnrecoverableValidationError):
+                raise e
+            raise UnrecoverableValidationError("Unhandled exception:", e) from e
+
+    return wrapper
+
+
+def _add_issue(issue_type: IssueType, issue: Issue):
+    """Add an Issue to the selected type for the current validation."""
+    context = current_context
+    if context.validation_name not in get_issues():
+        get_issues()[context.validation_name] = {
+            IssueType.PROBLEM.value: [],
+            IssueType.WARNING.value: [],
+        }
+    get_issues()[context.validation_name][issue_type.value].append(issue)
+
+
+def fail(
+    template: str, subjects: List[str] = None, resources: List[str] = None
+) -> None:
+    """Fail the validation."""
+    if template not in _issue_templates:
+        raise UnrecoverableValidationError(f"Unable to find issue template {template}")
```

Comment:
This would make sense to us, but for a customer running this tool, is there a better way to fail here? Will other validations run if we throw unrecoverable error here?

Reply:
No this is an immediate error only for us.  The customer should not see this.  UnrecoverableValidationError means it stops.